Igh covariance

globusharris opened the pull request.

Commit:
refactored utility tests

Commit:
Covariance fixes and testing

Commit:
moved file from R folder to tests

Commit:
updating documentation

Commit:
updated docs

Commit:
Merge branch 'IGH_covariance' of https://github.com/privacytoolsproject/PSI-Library into IGH_covariance

Commit:
Refactoring covariance to reflect input validation changes

Commit:
Refactored censordata to work w multiple range inputs

Commit:
Refactored existing covariance tests

Commit:
Added covariance sensitivity docs

Commit:
Added sensitivity special case for covariance w intercept

Commit:
Added validation checks to censorData

Commit:
Ammended censorData docs

Commit:
adding doc files

Commit:
Merge branch 'develop' of https://github.com/privacytoolsproject/PSI-Library into IGH_covariance

globusharris commented:
This resolves issues #62, #71, #75, 78, #89.

ctcovington commented on extra_docs/sensitivities/covariance_sensitivity.tex:
```diff
@@ -0,0 +1,255 @@
+\begin{document}
+
+\begin{definition}
+Let $X$ and $Y$ be two column vectors of length n. Then the sample covariance between $X$ and $Y$, $\cov[X,Y]$, is defined to be 
```
In my opinion, it might be worth changing the names of X and Y to avoid using X as both a vector (here) and a matrix (in definition 2).

globusharris replied:
Fixed.

ctcovington commented on extra_docs/sensitivities/covariance_sensitivity.tex:
```diff
@@ -0,0 +1,255 @@
+X = 
+\begin{bmatrix}
+X_1 & X_2 & \cdots & X_m
+\end{bmatrix}^T,
```
I think transposing here is consistent with the definition of the covariance matrix below, I'm just confused as to why we are transposing in the first place. It seems to me like we define X_i as column vectors and then treat them as rows for the rest of the document.

globusharris replied:
I defined them as column vectors since that's how they are implemented in the library.

ctcovington commented on extra_docs/sensitivities/covariance_sensitivity.tex:
```diff
@@ -0,0 +1,255 @@
+\end{corollary}
+
+\begin{theorem}
+Let $F(X)$ be the covariance matrix of $X$ without the normalization factor of $n-1$. Let $M_i$ be a maximum bound on $x_i \in X_i$, and let $m_i$ be a minimum bound on $x_i \in X_i$. Then each entry $f_{ij}$ of this matrix has sensitivity bounded above by
```
I have a related question to what I said on the variance sensitivity doc. What if X' has one fewer (rather than one more) observation than X? This might be covered later when we `consider the sensitivity of X'`, I just don't fully understand the framing of this as sensitivity of a random variable rather than of a function.

globusharris replied:
I updated docs to say sensitivity of f(X').

Shoeboxam committed:
fix /ref's in covariance_sensitivity.pdf

Shoeboxam commented on R/statistic-covariance.R:
```diff
@@ -30,77 +8,56 @@ coefficientRelease <- function(formula, release, n) {
+    diffs <- sapply(rng, diff)
+    if (intercept) { diffs <- c(1, diffs) }
+    sensitivity <- c()
+    c <- 2*(n-1)/n
```
Overwriting `c` can lead to unexpected behavior.
Also, why the special-casing on the intercept? If a user augments a column of ones, then the bounds will be (1, 1), giving zero sensitivity as expected. This is also hardcoding the location of the intercept column. Providing a feature for producing what a covariance matrix would look like if you were to augment X with a column of ones seems out of scope for the covariance statistic?

globusharris replied:
This is a feature that has existed since before I started working on the code; when the user initializes the statistic, they input a Boolean value for `intercept` and if it is `TRUE`, then the intercept column is always added.

So I think there are two questions here:
1. Is the original implementation, which lets users choose if they want an intercept column added or not, out of the scope of the statistic?
2. Why the special casing on the intercept?

For the second question, you're right that the special casing is unnecessary/incorrect. I'm not sure why I append a 1 to the diffs, when it should be a 0 anyhow, and that means that the casing later on is also unnecessary.

If we do want to keep the intercept option in the initialization, some sort of special-casing somewhere will be necessary (at a minimum, adding to the diff ranges). But it could be moved outside of the sensitivity calculation, which I'll do at a minimum. James might be the person to ask about the first question.

Good catch about the `c` assignment; it's my de facto variable for constants and I didn't think about how it would interact with `c()`.

globusharris replied:
I removed special casing and changed `c` assignment. Question 1 still remains unanswered.

Commit:
Removed excessive intercept casing on intercept calc, changed variable name from 'c'.

Commit:
Merge branch 'IGH_covariance' of https://github.com/privacytoolsproject/PSI-Library into IGH_covariance

Commit:
Resolved merge conflicts with develop

Merge branch 'develop' of https://github.com/privacytoolsproject/PSI-Library into IGH_covariance

# Conflicts:
#	man/checkRange.Rd
#	tests/testthat/test-utilities.R

Commit:
updated covariance docs

Commit:
Reran documentation
